src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,169 @@
+        if (index < 0 || index > currentSlot) {
+            throw new ArrayListIndexOutOfBoundsException("Index " + index
+                    + " out of bounds ");
+        }
+        return array[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        if (index < 0 || index > currentSlot) {
+            throw new ArrayListIndexOutOfBoundsException("Index " + index
+                    + " out of bounds ");
+        }
+        array[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        if (index < 0 || index >= size()) {
+            throw new ArrayListIndexOutOfBoundsException("Index " + index
+                    + " out of bounds");
+        }
+        T date = array[index];
+        for (int i = index; i < size(); i++) {
+            if (index == currentSlot) {
+                array[index] = null;
+                currentSlot--;
+                return date;
+            }
+            array[i] = array[i + 1];
+        }
+        currentSlot--;
+        array[size()] = null;
+        return date;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        if (isEmpty()) {
+            throw new NoSuchElementException("Found element doesn't exist");
+        }
+        int indexFoundElement = -1;
+        search:
+        for (int i = 0; i < size(); i++) {
+            if (array[i] == null || element == null) {
+                indexFoundElement = i;
+                break search;
+            }
+            if (array[i] == element || array[i].equals(element)) {
+                indexFoundElement = i;
+                break search;
+            }
+        }
+        if (indexFoundElement == -1) {
+            throw new NoSuchElementException("Found element doesn't exist");
+        } else {
+            for (int i = indexFoundElement; i < size(); i++) {
+                if (indexFoundElement == size() - 1) {
+                    array[indexFoundElement] = null;
+                    currentSlot--;
+                    return element;
+                }
+                array[i] = array[i + 1];
+            }
+            array[currentSlot] = null;
+            currentSlot--;
+        }
+        return element;
     }
 
     @Override
     public int size() {
-        return 0;
+        if (array == null) {
+            return 0;
+        }
+        return currentSlot + 1;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        if (currentSlot > -1) {
+            return false;
+        }
+        return true;
+    }
+
+    private void checkAndIncreaseArrayLength() {
+        T[] tempArray = null;
+        if (array[array.length - 1] != null) {
+            tempArray = (T[]) new Object[(int) (array.length * ARRAY_INCREMENT_VALUE)];
+            for (int i = 0; i < array.length; i++) {
+                tempArray[i] = array[i];
+            }
+            array = tempArray;
+        }
+    }
```

Comment:
you should be able to just call `increaseArrayLength(size() + 1)` here or something like this